File: client/src/main/java/net/treset/minecraftlauncher/creation/InstanceCreator.java
package net.treset.minecraftlauncher.creation;

import net.treset.mc_version_loader.launcher.*;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.util.CreationStatus;
import net.treset.minecraftlauncher.util.exception.ComponentCreationException;
import net.treset.minecraftlauncher.util.string.PatternString;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class InstanceCreator extends GenericComponentCreator {
    private static final Logger LOGGER = LogManager.getLogger(InstanceCreator.class);

    private final List<PatternString> ignoredFiles;
    private final List<LauncherLaunchArgument> jvmArguments;
    private final List<LauncherFeature> features;
    private final ModsCreator modsCreator;
    private final OptionsCreator optionsCreator;
    private final ResourcepackCreator resourcepackCreator;
    private final SavesCreator savesCreator;
    private final VersionCreator versionCreator;

    public InstanceCreator(String name, Map<String, LauncherManifestType> typeConversion, LauncherManifest componentsManifest, List<PatternString> ignoredFiles, List<LauncherLaunchArgument> jvmArguments, List<LauncherFeature> features, ModsCreator modsCreator, OptionsCreator optionsCreator, ResourcepackCreator resourcepackCreator, SavesCreator savesCreator, VersionCreator versionCreator) {
        super(LauncherManifestType.INSTANCE_COMPONENT, null, null, name, typeConversion, LauncherApplication.config.INSTANCE_DEFAULT_INCLUDED_FILES, LauncherApplication.config.INSTANCE_DEFAULT_DETAILS, componentsManifest);
        this.ignoredFiles = ignoredFiles;
        this.jvmArguments = jvmArguments;
        this.features = features;
        this.modsCreator = modsCreator;
        this.optionsCreator = optionsCreator;
        this.resourcepackCreator = resourcepackCreator;
        this.savesCreator = savesCreator;
        this.versionCreator = versionCreator;
        setDefaultStatus(new CreationStatus(CreationStatus.DownloadStep.STARTING, null));
    }


    @Override
    public String createComponent() throws ComponentCreationException {
        String result = super.createComponent();

        if(result == null || getNewManifest() == null) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create instance component: invalid data");
        }

        ArrayList<LauncherFeature> features = new ArrayList<>(this.features);
        features.addAll(LauncherApplication.config.INSTANCE_DEFAULT_FEATURES);

        ArrayList<PatternString> ignoredFiles = new ArrayList<>(this.ignoredFiles);
        ignoredFiles.addAll(LauncherApplication.config.INSTANCE_DEFAULT_IGNORED_FILES);

        ArrayList<LauncherLaunchArgument> jvmArguments = new ArrayList<>(this.jvmArguments);
        jvmArguments.addAll(LauncherApplication.config.INSTANCE_DEFAULT_JVM_ARGUMENTS);

        LauncherInstanceDetails details = new LauncherInstanceDetails(
                features,
                ignoredFiles.stream().map(PatternString::get).toList(),
                jvmArguments,
                null, null, null, null, null);

        try {
            if (modsCreator != null) {
                details.setModsComponent(modsCreator.getId());
            }
            details.setOptionsComponent(optionsCreator.getId());
            details.setResourcepacksComponent(resourcepackCreator.getId());
            details.setSavesComponent(savesCreator.getId());
            details.setVersionComponent(versionCreator.getId());
        } catch (ComponentCreationException e) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create instance: Error creating components", e);
        }

        setStatus(new CreationStatus(CreationStatus.DownloadStep.FINISHING, null));

        try {
            details.writeToFile(getNewManifest().getDirectory() + getNewManifest().getDetails());
        } catch (IOException e) {
            attemptCleanup();
            throw new ComponentCreationException("Failed to create instance component: failed to write details to file", e);
        }

        LOGGER.debug("Created instance component: id={}", getNewManifest().getId());
        return result;
    }

    @Override
    public String inheritComponent() throws ComponentCreationException {
        attemptCleanup();
        throw new ComponentCreationException("Unable to inherit instance component: unsupported");
    }

    @Override
    public String useComponent() throws ComponentCreationException {
        attemptCleanup();
        throw new ComponentCreationException("Unable to use instance component: unsupported");
    }

    @Override
    protected boolean attemptCleanup() {
        boolean success = super.attemptCleanup();
        success &= optionsCreator.attemptCleanup();
        success &= modsCreator == null || modsCreator.attemptCleanup();
        success &= savesCreator.attemptCleanup();
        success &= versionCreator.attemptCleanup();
        success &= resourcepackCreator.attemptCleanup();
        LOGGER.debug("Attempted cleanup of instance component: success={}", success);
        return success;
    }

    @Override
    public void setStatusCallback(Consumer<CreationStatus> statusCallback) {
        super.setStatusCallback(statusCallback);
        optionsCreator.setStatusCallback(statusCallback);
        if(modsCreator != null) {
            modsCreator.setStatusCallback(statusCallback);
        }
        savesCreator.setStatusCallback(statusCallback);
        versionCreator.setStatusCallback(statusCallback);
        resourcepackCreator.setStatusCallback(statusCallback);
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/creation/InstanceCreator.kt
package net.treset.treelauncher.backend.creation

import io.github.oshai.kotlinlogging.KotlinLogging
import net.treset.treelauncher.backend.config.appConfig
import net.treset.treelauncher.backend.data.LauncherFeature
import net.treset.treelauncher.backend.data.LauncherInstanceDetails
import net.treset.treelauncher.backend.data.LauncherLaunchArgument
import net.treset.treelauncher.backend.data.manifest.LauncherManifestType
import net.treset.treelauncher.backend.data.manifest.ParentManifest
import net.treset.treelauncher.backend.util.CreationStatus
import net.treset.treelauncher.backend.util.exception.ComponentCreationException
import net.treset.treelauncher.backend.util.file.LauncherFile
import net.treset.treelauncher.backend.util.string.PatternString
import java.io.IOException

class InstanceCreator(
    name: String?,
    typeConversion: Map<String, LauncherManifestType>?,
    componentsManifest: ParentManifest?,
    private val ignoredFiles: List<PatternString>,
    private val jvmArguments: List<LauncherLaunchArgument>,
    private val features: List<LauncherFeature>,
    private val modsCreator: ModsCreator?,
    private val optionsCreator: OptionsCreator,
    private val resourcepackCreator: ResourcepackCreator,
    private val savesCreator: SavesCreator,
    private val versionCreator: VersionCreator
) : GenericComponentCreator(
    LauncherManifestType.INSTANCE_COMPONENT,
    null,
    null,
    name,
    typeConversion,
    appConfig().instanceDefaultIncludedFiles,
    appConfig().instanceDefaultDetails,
    componentsManifest
) {

    init {
        defaultStatus = CreationStatus(CreationStatus.DownloadStep.STARTING, null)
    }

    @Throws(ComponentCreationException::class)
    override fun createComponent(): String {
        val result = super.createComponent()
        newManifest?.let { newManifest ->
            val features: ArrayList<LauncherFeature> = ArrayList(features)
            features.addAll(appConfig().instanceDefaultFeatures)
            val ignoredFiles: ArrayList<PatternString> = ArrayList(ignoredFiles)
            ignoredFiles.addAll(appConfig().instanceDefaultIgnoredFiles)
            val jvmArguments: ArrayList<LauncherLaunchArgument> = ArrayList(
                jvmArguments
            )
            jvmArguments.addAll(appConfig().instanceDefaultJvmArguments)
            val details = try {
                LauncherInstanceDetails(
                    features,
                    ignoredFiles.stream().map(PatternString::get).toList(),
                    jvmArguments,
                    modsCreator?.id,
                    optionsCreator.id,
                    resourcepackCreator.id,
                    savesCreator.id,
                    versionCreator.id
                )
            } catch (e: ComponentCreationException) {
                attemptCleanup()
                throw ComponentCreationException("Failed to create instance: Error creating components", e)
            }
            setStatus(CreationStatus(CreationStatus.DownloadStep.FINISHING, null))
            try {
                LauncherFile.of(newManifest.directory, newManifest.details).write(details)
            } catch (e: IOException) {
                attemptCleanup()
                throw ComponentCreationException(
                    "Failed to create instance component: failed to write details to file",
                    e
                )
            }
            LOGGER.debug { "${"Created instance component: id={}"} ${newManifest.id}" }
        }?: run {
            attemptCleanup()
            throw ComponentCreationException("Failed to create instance component: invalid data")
        }
        return result
    }

    @Throws(ComponentCreationException::class)
    override fun inheritComponent(): String {
        attemptCleanup()
        throw ComponentCreationException("Unable to inherit instance component: unsupported")
    }

    @Throws(ComponentCreationException::class)
    override fun useComponent(): String {
        attemptCleanup()
        throw ComponentCreationException("Unable to use instance component: unsupported")
    }

    override fun attemptCleanup(): Boolean {
        var success = super.attemptCleanup()
        success = success and optionsCreator.attemptCleanup()
        success = success and (modsCreator == null || modsCreator.attemptCleanup())
        success = success and savesCreator.attemptCleanup()
        success = success and versionCreator.attemptCleanup()
        success = success and resourcepackCreator.attemptCleanup()
        LOGGER.debug { "${"Attempted cleanup of instance component: success={}"} $success" }
        return success
    }

    override var statusCallback: (CreationStatus) -> Unit
        get() = super.statusCallback
        set(statusCallback) {
            super.statusCallback = statusCallback
            optionsCreator.statusCallback = statusCallback
            modsCreator?.statusCallback = statusCallback
            savesCreator.statusCallback = statusCallback
            versionCreator.statusCallback = statusCallback
            resourcepackCreator.statusCallback = statusCallback
        }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
    }
}
